From 64cc243c33ae10e02ceee6c19a5dc2f6081043d4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Pierret=20=28fepitre=29?=
 <frederic.pierret@qubes-os.org>
Date: Sat, 16 May 2020 11:44:53 +0200
Subject: [PATCH 5/8] Qubes decoration: handle guivm windows prefix

Authors:
@fepitre

Original work done in:
- QubesOS/qubes-desktop-linux-xfce4-xfwm4
---
 src/client.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/src/client.c b/src/client.c
index efb90c361..7d67d2353 100644
--- a/src/client.c
+++ b/src/client.c
@@ -291,12 +291,32 @@ getQubesVmName (Client *c)
                     c->qubes_vmname = g_strdup (_("[ERROR Reading VM name?!] "));
                 }
             } else {
-                c->qubes_vmname = g_strdup("[Dom0] ");
+                // In case of GuiVM get qubes_vmname from root window if set
+                if ((XGetWindowProperty (display_info->dpy, XDefaultRootWindow(display_info->dpy), atom_vmname, 0L, 0L,
+                                FALSE, AnyPropertyType, &actual_type, &actual_format, &nitems,
+                                &bytes_left, (unsigned char **) &data) == Success))
+                {
+                    if (bytes_left > 0) {
+                        if ((XGetWindowProperty (display_info->dpy, XDefaultRootWindow(display_info->dpy), atom_vmname, 0L, bytes_left,
+                                        FALSE, AnyPropertyType, &actual_type, &actual_format, &nitems,
+                                        &bytes_left, (unsigned char **) &data) == Success))
+                        {
+                            c->qubes_vmname = g_strdup_printf (_("[%s] "), data);
+                        } else {
+                            c->qubes_vmname = g_strdup (_("[ERROR Reading VM name?!] "));
+                        }
+                    } else {
+                        c->qubes_vmname = g_strdup("[Dom0] ");
+                    }
+                } else {
+                    c->qubes_vmname = g_strdup("[Dom0] ");
+                }
             }
         } else {
             c->qubes_vmname = g_strdup("[Dom0] ");
         }
     }
+    TRACE ("qubes_vmname \"%s\"", c->qubes_vmname);
 }
 
 static gchar*
-- 
2.48.1

